internal/impl/sftp/input.go

```diff
@@ -174,11 +174,62 @@ func newSFTPReaderFromParsed(conf *service.ParsedConfig, mgr *service.Resources)
 }
 
 func (s *sftpReader) Connect(ctx context.Context) (err error) {
+	file, nextPath, skip, err := s.seekNextPath(ctx)
+	if err != nil {
+		return err
+	}
+	if skip {
+		return nil
+	}
+
+	details := service.NewScannerSourceDetails()
+	details.SetName(nextPath)
+	if s.scanner, err = s.scannerCtor.Create(file, func(ctx context.Context, aErr error) (outErr error) {
+		_ = s.pathProvider.Ack(ctx, nextPath, aErr)
+		if aErr != nil {
+			s.log.Errorf("skipping delete on finish: %s", aErr)
+			return nil
+		}
+		if s.deleteOnFinish {
+			s.scannerMut.Lock()
+			client := s.client
+			if client == nil {
+				if client, outErr = s.creds.GetClient(s.mgr.FS(), s.address); outErr != nil {
+					outErr = fmt.Errorf("obtain private client: %w", outErr)
+				}
+				defer func() {
+					_ = client.Close()
+				}()
+			}
+			if outErr == nil {
+				if outErr = client.Remove(nextPath); outErr != nil {
+					outErr = fmt.Errorf("remove %v: %w", nextPath, outErr)
+				}
+			}
+			s.scannerMut.Unlock()
```

Comment:
since we always return after this block we could defer this right? I just get worried if the unlock is not right after.